// capture/ElecInterface.js
/**
 * ElecInterface
 *
 * @description :: ElecInterface
 * @author      :: 石建森
 */

// Import dependencies
const Convention = require('./Convention.js');
const STATUS = Convention.STATUS;
const DATA_TYPE = Convention.DATA_TYPE;
const config = require('./Config.js');
const CrawlAutomata = require('../CrawlAutomata.js');
const elecParser = require('ElecParser.js');
const cheerio = require('cheerio');

/**
 * Some custom functions, codes, or some modules
 * NOTE: This is a node module, variables below are 'static'
 */

/**
 * Capture info
 * getSessionID: ['http://elec.xmu.edu.cn/PdmlWebSetup/Pages/SMSMain.aspx', 'get'],
 * click: ['http://elec.xmu.edu.cn/PdmlWebSetup/Pages/SMSMain.aspx', 'post'],
 * inquiry: ['http://elec.xmu.edu.cn/PdmlWebSetup/Pages/SMSMain.aspx', 'post']
 */

/**
 * @param {Object} option may contain username and password
 * @param {function} callback callback function. Prototype: function({result})
 */
module.exports = (option, callback) => {
  // Do something here
  // Capture and process data
  var automata = new CrawlAutomata();
  function onError(status) {
    callback({
      status: status || STATUS.SERVER_FAILURE // STATUS.OK == 0
    });
    // console.log('Error! ' + status);
  }
  function onSuccess(data) {
    callback({
      status: STATUS.OK,
      dataType: DATA_TYPE.JSON,
      data: data
    });
  }

  var _dormitory = null; // An instance of Use
  automata.states = {
    getSessionID: {
      url: 'http://elec.xmu.edu.cn/PdmlWebSetup/Pages/SMSMain.aspx',
      method: 'get',
      callback: ((err, res) => {
        if(err || !res.ok) {
          var status = STATUS.SERVER_FAILURE;
          if(err.timeout) {
            sails.log.info('[ Elec ] Get session id timeout. ' + err);
            status = STATUS.TIMEOUT;
          } else {
            sails.log.info('[ Elec ] Get session id failed. ' + err);
          }
          onError(status);
        } else {
          var $ = cheerio.load(res.text);
          automata.states.click.data.__VIEWSTATE = $('#__VIEWSTATE').val();
          automata.states.click.data.__EVENTVALIDATION = $('#__EVENTVALIDATION').val();
          automata.states.click.data.dxgvSubInfo$CallbackState = $('#dxgvSubInfo_CallbackState').val();
          automata.states.click.data.dxgvElec$CallbackState = $('#dxgvElec_CallbackState').val();
          // console.log(VIEWSTATE, EVENTVALIDATION);
          sails.log.info('[ Elec ] Session ID get');
          sails.log.silly(res.text);
          automata.transit('click');
        }
      })
    },
    click: {
      url: 'http://elec.xmu.edu.cn/PdmlWebSetup/Pages/SMSMain.aspx',
      method: 'post',
      type: 'form',
      data: {
        '__VIEWSTATE': 'rua',
        '__EVENTVALIDATION': 'rua',
        'dxgvSubInfo$CallbackState': 'rua',
        'dxgvElec$CallbackState': 'rua',
        'drxiaoqu': option.drxiaoqu,
        'drlou': '',
        'txtRoomid': '',
        'dxdateStart_Raw': option.dxdateStart_Raw,
        'dxdateEnd_Raw': option.dxdateEnd_Raw,
      },
      callback: ((err, res) => {
        if(err || !res.ok) {
          var status = STATUS.SERVER_FAILURE;
          if(err.timeout) {
            sails.log.info('[ Elec ] Get clicky stuff timeout. ' + err);
            status = STATUS.TIMEOUT;
          } else {
            sails.log.info('[ Elec ] Get clicky stuff failed. ' + err);
          }
          onError(status);
        } else {
          var $ = cheerio.load(res.text);
          automata.states.inquiry.data.__VIEWSTATE = $('#__VIEWSTATE').val();
          automata.states.inquiry.data.__EVENTVALIDATION = $('#__EVENTVALIDATION').val();
          automata.states.inquiry.data.dxgvSubInfo$CallbackState = $('#dxgvSubInfo_CallbackState').val();
          automata.states.inquiry.data.dxgvElec$CallbackState = $('#dxgvElec_CallbackState').val();
          // console.log(VIEWSTATE, EVENTVALIDATION);
          sails.log.info('[ Elec ] Get clicky stuff success');
          sails.log.silly(res.text);
          automata.transit('inquiry');
        }
      })
    },
    inquiry: {
      url: 'http://elec.xmu.edu.cn/PdmlWebSetup/Pages/SMSMain.aspx',
      method: 'post',
      type: 'form',
      data: {
        '__VIEWSTATE': 'rua',
        '__EVENTVALIDATION': 'rua',
        'dxgvSubInfo$CallbackState': 'rua',
        'dxgvElec$CallbackState': 'rua',
        'drxiaoqu': option.drxiaoqu,
        'drlou': option.drlou,
        'txtRoomid': option.txtRoomid,
        'dxdateStart_Raw': option.dxdateStart_Raw,
        'dxdateEnd_Raw': option.dxdateEnd_Raw,
      },
      callback: ((err, res) => {
        if(err || !res.ok) {
          var status = STATUS.SERVER_FAILURE;
          if(err.timeout) {
            sails.log.info('[ Elec ] Inquiry timeout. ' + err);
            status = STATUS.TIMEOUT;
          } else {
            sails.log.info('[ Elec ] Inquiry failed. ' + err);
          }
          // console.log(err);
        } else /*if(res.text.match('Success'))*/ {
          sails.log.info('[ Elec ] Inquiry success. Domitory:', option.drxiaoqu, option.drlou, option.txtRoomid);
          onSuccess(elecParser(res.text));
        }
      })
    },
  };
  automata.start('getSessionID');
};
